pkg/cloudclient/aws/private.go

```diff
@@ -1099,3 +1110,110 @@ func encodeAWSMachineProviderSpec(awsProviderSpec *machinev1beta1.AWSMachineProv
 		Raw: buffer.Bytes(),
 	}, nil
 }
+
+// This code should run on non-CPMS clusters
+func removeLoadBalancerMachineSet(ctx context.Context, kclient k8s.Client, lbName string, masterList *machinev1beta1.MachineList) error {
+	err := removeAWSLBFromMasterMachines(kclient, lbName, masterList)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func removeLoadBalancerCPMS(ctx context.Context, kclient k8s.Client, lbName string, cpms *machinev1.ControlPlaneMachineSet) error {
+	cpmsPatch := k8s.MergeFrom(cpms.DeepCopy())
+	rawExtension := cpms.Spec.Template.OpenShiftMachineV1Beta1Machine.Spec.ProviderSpec.Value
+	spec, err := baseutils.ConvertFromRawExtension[machinev1beta1.AWSMachineProviderConfig](rawExtension)
+	if err != nil {
+		return err
+	}
+	var remainingLoadBalancers []machinev1beta1.LoadBalancerReference
+	for _, lb := range spec.LoadBalancers {
+		if lb.Name == lbName {
+			log.Info("Removing loadbalancer %s from CPMs", "load-balancer-name", lbName)
+		} else {
+			log.Info("Keeping loadbalancer %s from CPMs", "load-balancer-name", lb.Name)
+			remainingLoadBalancers = append(remainingLoadBalancers, lb)
+		}
+	}
+	spec.LoadBalancers = remainingLoadBalancers
+	extension, err := baseutils.ConvertToRawBytes(spec)
+	if err != nil {
+		return err
+	}
+	cpms.Spec.Template.OpenShiftMachineV1Beta1Machine.Spec.ProviderSpec.Value.Raw = extension
+	err = kclient.Patch(ctx, cpms, cpmsPatch)
+	if err != nil {
+		return fmt.Errorf("could not update CPMS: %v", err)
+	}
+	return nil
+}
+
+func getLoadBalancerRemovalFunc(ctx context.Context, kclient k8s.Client, masterList *machinev1beta1.MachineList, cpms *machinev1.ControlPlaneMachineSet) func(string) error {
+	if cpms.Spec.State == machinev1.ControlPlaneMachineSetStateActive {
+		return func(lbName string) error {
+			log.Info("Removing active CPMS")
+			err := baseutils.DeleteCPMS(ctx, kclient, cpms)
+			if err != nil {
+				log.Error(err, "failed to delete CPMS")
+				return err
+			}
+			// It seems this has to actually do both - the machine api is not
+			// able to remove machines that still reference a LB that does not
+			// exist:
+			// E1117 14:42:49.349992       1 controller.go:255] fbergman-14568-tn7t8-master-0: failed to delete machine: fbergman-14568-tn7t8-master-0: reconciler failed to Delete machine: failed to updated update load balancers: LoadBalancerNotFound: Load balancers '[fbergman-14568-tn7t8-ext]' not found
+			log.Info("Updating controlplane machines")
+			err = removeLoadBalancerMachineSet(ctx, kclient, lbName, masterList)
+			if err != nil {
+				log.Error(err, "failed to remove load balancer from machines")
+				return err
+			}
+			go func() {
+				maxRetries := 5
+				for {
```

Comment:
```suggestion
				for maxRetries := 0; maxRetries < 5; maxRetries++ {
```

Reply:
I kept the infinite loop as that makes it easier to handle the 'it did not succeed -> print a log' case.
If I use a normal loop I would have to name it and break it to differentiate the cases:
* Broke the loop because success
* Broke the loop because all attempts expired